test: Unit tests for taproot/tapscript coverage in `interpreter.cpp`

david-bakin opened the pull request:
[Unit tests are not complete to resolve #23279 - _yet_.  But those additional tests - see this space for future developments! - will not change the code presented here.]

This PR provides unit tests for Taproot functionality in `script/interpreter.cpp`, #23279.

Goal is to show correctness vs specifications in BIPs 341/342, and to improve code coverage in Taproot functionality there.

Intent is to unit test completely without changing `interpreter.cpp` itself to provide visibility into static functions/private methods.  Thus some of the tests must be "indirect", testing the unit-under-test by probing a visible caller.  This makes those specific tests somewhat fragile as they depend on code paths in the caller, which may change in the future.  However, such tests are written so they will fail if the unit-under-test isn't actually called.  This is noted in the code of those specific tests.

No changes to Bitcoin Core are made - just unit tests added.  The Boost Test framework file `boost/test/execution_monitor.hpp` was added to the lint whitelist - this is acceptable because it is a documented public header of the Boost Test framework, which is already part of the project.

- [x] `EvalChecksigTapscript`
- [x] `HandleMissingData`
- [x] `SignatureHashSchnorr` - now with full _path_ coverage
- [x] `GenericTransactionSignatureChecker<T>::VerifySchnorrSignature`
- [x] `GenericTransactionSignatureChecker<T>::CheckSchnorrSignature`
- [x] `ComputeTaprootMerkleRoot`
- [ ] `SigVersion::TAPSCRIPT` code paths in `ExecuteWitnessScript`
- [x] Taproot code paths in `VerifyWitnessProgram`
- [ ] Tapscript code paths in `VerifyWitnessProgram`

**TODO:** Full Tapscript (BIP-342) unit tests yet to be completed.
**TODO:** Does anything need to be done with `libconsensus`? Please advise!

2 commits:

1. Add new utility functions to `src/test/util` that are not only used in these tests but may also be useful for future unit tests.  Also, refactor some existing utility functions from current tests (mainly, `script_tests.cpp`) into `src/test/util` so they can be more widely used too.  All of these new test utility functions have unit tests themselves, in `src/test/test_utils_tests.cpp`.
1. All the tests for Taproot/Tapscript functionality requested by #23279 in a new file `src/test/script_tapscript_tests.cpp`.  Code coverage runs show 100% coverage of lines and near 100% coverage of branches.

**DEATH TRAP AHEAD!:** These unit tests include "death tests" (name comes from Googletest framework): Tests that check that an `assert` actually fails.  These are handled with a Boost `execution_monitor` (mentioned above) _but_ the asserts _still issue messages_ even when correctly trapped by the Boost Test framework.  You'll see the following in the logs:

```
gitpod /workspace/bitcoin/src/test (23279-taproot-unit-tests) $ ./test_bitcoin 
Running 527 test cases...
test_bitcoin: script/interpreter.cpp:1495: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `false' failed.
test_bitcoin: script/interpreter.cpp:1497: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `in_pos < tx_to.vin.size()' failed.
test_bitcoin: script/interpreter.cpp:1528: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_annex_init' failed.
test_bitcoin: script/interpreter.cpp:1556: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_tapleaf_hash_init' failed.
test_bitcoin: script/interpreter.cpp:1559: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_codeseparator_pos_init' failed.
test_bitcoin: script/interpreter.cpp:1469: bool HandleMissingData(MissingDataBehavior): Assertion `!"Missing data"' failed.
test_bitcoin: script/interpreter.cpp:1474: bool HandleMissingData(MissingDataBehavior): Assertion `!"Unknown MissingDataBehavior value"' failed.

*** No errors detected
gitpod /workspace/bitcoin/src/test (23279-taproot-unit-tests) $ 
```

These specific `Assertion ... failed` messages are _expected_.  The tests triggering them properly _pass_ iff the `assert` (and no other) is _triggered_.

-----

Coverage results 2022-06-05 showing Taproot coverage (Tapscript tests not yet complete) when _only running_ `script_tapscript_tests.cpp`:

|  |   |
|--:|:--|
|`EvalChecksigTapscript`|<img alt="EvalChecksigTapscript" align="left" width="550" src="https://user-images.githubusercontent.com/4162948/172067379-4815fd13-1eca-4fdc-afb3-ab321f644ca6.png">|
|`HandleMissingData`|<img alt="HandleMissingData coverage" align="left" width="550" src="https://user-images.githubusercontent.com/4162948/172066800-7ad85065-448c-4300-8f5f-bbdaeb5bc885.png">|
|`SignatureHashSchnorr`|<img alt="SignatureHashSchnorr coverage" align="left" width="550" src="https://user-images.githubusercontent.com/4162948/172067034-c8d65908-bf41-4420-8a5d-5c023f85824a.png">|
|`GenericTransaction-`<br>`SignatureChecker<T>::`<br>`CheckSchnorrSignature`|<img alt="CheckSchnorrSignature" align="left" width="550" src="https://user-images.githubusercontent.com/4162948/172067232-b58ea75a-9df7-4f02-b291-c1adda4792e8.png">|
|`ComputeTaprootMerkleRoot`|<img alt="ComputeTaprootMerkleRoot" align="left" width="550" src="https://user-images.githubusercontent.com/4162948/172067270-d15ac048-6001-4a8f-b0a7-f62b76cf584b.png">|
|`VerifyWitnessProgram`<br>(Taproot only)|<img alt="VerifyWitnessProgram" align="left" width="550" src="https://user-images.githubusercontent.com/4162948/172067308-c23ccaaa-dbee-43ea-b2df-cbee9ba375ed.png">|

david-bakin commented:
(`Win64 native [msvc]` failed above in functional test `feature_index_prune.py` - this test has been reported recently as flaky, see #25031.)

jonatack commented:
Squashed and rebased this branch to master and seeing this when running the tests:
```
$  ./src/test/test_bitcoin -t script_tapscript_tests
Running 12 test cases...
test_bitcoin: script/interpreter.cpp:1495: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `false' failed.
test_bitcoin: script/interpreter.cpp:1497: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `in_pos < tx_to.vin.size()' failed.
test_bitcoin: script/interpreter.cpp:1528: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_annex_init' failed.
test_bitcoin: script/interpreter.cpp:1556: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_tapleaf_hash_init' failed.
test_bitcoin: script/interpreter.cpp:1559: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_codeseparator_pos_init' failed.
test_bitcoin: script/interpreter.cpp:1469: bool HandleMissingData(MissingDataBehavior): Assertion `!"Missing data"' failed.
test_bitcoin: script/interpreter.cpp:1474: bool HandleMissingData(MissingDataBehavior): Assertion `!"Unknown MissingDataBehavior value"' failed.
```

david-bakin commented:
> Squashed and rebased this branch to master and seeing this when running the tests:
> 
> ```
> $  ./src/test/test_bitcoin -t script_tapscript_tests
> Running 12 test cases...
> test_bitcoin: script/interpreter.cpp:1495: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `false' failed.


Yes - that's actually expected.  Those are death tests - testing asserts.  The Boost Test framework will print that the assertion failed but then it is properly caught with an `ExecutionMonitor` so the test actually succeeds.

The reason for these tests is that BIP-341 specifies several errors in the signature computation step but the routine in `interpreter.cpp` implements _some_ of those checks with asserts.  (Other checks are implemented with a return code, and a few checks actually choose _at runtime_ whether to assert or return an error code! I don't know why it was written this way.)

Do you have any suggestions on how I can improve the documentation or the test to avoid confusion such as this?  Perhaps I could use `BOOST_TEST_MESSAGE` to emit a message: "These following assertions are _expected_"??

jonatack commented:
Oh ok, I hadn't looked at the code yet (the fixup and merge commits really need to be squashed to encourage people to look at it) but indeed there is a large comment. Agree, logging a message to explain would be helpful.

david-bakin commented:
I will squash now (actually, in the morning ...) - I was waiting until I finished the last couple of tasks but no problem doing it now.  Thanks for the tip.

jonatack commented:
I tend to run individual tests with `-l test_suite` if anything unusual occurs when  running it without a loglevel.

 In any case, `make check` runs it fine locally for me and it looks like you solved your CI issue.

jonatack commented on src/test/script_tapscript_tests.cpp:
```diff
@@ -0,0 +1,1546 @@
+                                  * output_hash_alternatives.size()
+                                  - 8 /* exclude SIGHASH_DEFAULT w/ SISHASH_ANYONECANPAY */;
+
+        BOOST_TEST_MESSAGE("Running " << nAlternatives << "scenarios");
```
Noticed this in the test output.
```suggestion
        BOOST_TEST_MESSAGE("Running " << nAlternatives << " scenarios");
```

david-bakin replied:
Fixed.

jamesob commented on src/test/script_tapscript_tests.cpp:
```diff
@@ -0,0 +1,1546 @@
+    }
+
+    // hex strings to byte vector
+    BOOST_TEST("0A"_hex == valtype(1, 0x0a));
```
Wow, TIL about user-defined literals in C++11.

So these are cool, but is the "xx"_hex stuff actually used in taproot-specific tests anywhere here? I can't find any usages, which would make that facility (and these tests) unnecessary if I'm reading right.

david-bakin replied:
I'm going to move all of these helpers off into homes in `src/test/util` where they can be used to make unit tests easier to write/read/maintain.  But reduce the set of them at the same time - there's some redundancy, as you've noticed.  (In a subsequent push to this PR.)

david-bakin replied:
And that's done now.

jamesob reviewed (commented):
Concept ACK

Nice job so far! These tests are well formulated. I like how declarative and readable they are. I'll review more of the substance once this exits draft state.

david-bakin commented:
> Edit: nit fixups signaled by `test/lint/lint-spelling.py`

Fixed, and I'll run `lint-spelling` on my own from now on.

> Edit 2: I did see this error with `make check`, unsure if it was with your latest push:
> 
> ```
> Running tests: from test/script_tapscript_tests.cpp
> cat: test/script_tapscript_tests.cpp: No such file or directory
> Missing an argument value for the parameter run_test in the argument 
> 
>   run_test
> ...
> ```

I do not see this when running `make check` or when running `test_bitcoin` either with or without `--run_test=script_tapscript_tests`.  Relevant section of my terminal spew when running `make check` locally:

```
...
Running tests: script_tapscript_tests from test/script_tapscript_tests.cpp
export TEST_LOGFILE=/workspace/bitcoin/src/$( echo test/script_tapscript_tests.cpp | grep -E -o "(wallet/test/.*\.cpp|test/.*\.cpp)" | /usr/bin/sed -e s/\.cpp/.log/ ) && \
test/test_bitcoin --catch_system_errors=no -l test_suite -t "$( cat test/script_tapscript_tests.cpp | grep -E "(BOOST_FIXTURE_TEST_SUITE\\(|BOOST_AUTO_TEST_SUITE\\()" | cut -d '(' -f 2 | cut -d ',' -f 1 | cut -d ')' -f 1 )" -- DEBUG_LOG_OUT > "$TEST_LOGFILE" 2>&1 || (cat "$TEST_LOGFILE" && false)
Running tests: script_tests from test/script_tests.cpp
...
```
And the CI checks all pass too.  Not sure how to proceed on this report.

david-bakin commented on src/test/script_tapscript_tests.cpp:
```diff
@@ -0,0 +1,1546 @@
+                                    Span<const unsigned char> pubkey,
+                                    SigVersion sigversion,
+                                    ScriptExecutionData& execdata,
+                                    ScriptError* serror = nullptr) const override
```
OK!

david-bakin commented:
Squashed now as requested.  Still 2 tests left to write: They're tough ones as both are for `static` methods well hidden by the accessible public method, and no easy hooks like polymorphic classes to mock/fake to get in there with.

david-bakin committed:
Add new test utility methods; break out some existing ones

- Add new test utility methods for use in various unit tests.
- Refactor some existing methods out of tests where they're currently
  embedded into the test util area.
- Ensure that the conversions of string name to/from ScriptError_t have
  all the defined enum values.
- And provide unit tests for the whole shebang.

david-bakin committed:
Unit tests for Tap{root,script} of `interpreter.cpp`, #23279

This commit provides unit tests for Taproot functionality in
`script/interpreter.cpp`, #23279.

Goal is to show correctness vs specifications in BIPs 341/342, and to
improve code coverage in Taproot functionality there.

Intent is to unit test completely without changing `interpreter.cpp`
itself to provide visibility into static functions/private methods.
Thus some of the tests must be "indirect", testing the unit-under-test
by probing a visible caller. This makes those specific somewhat fragile
as they depend on code paths in the caller, which may change in the
future.  However, such tests are written so they will fail if the
unit-under-test isn't actually called. This is noted in the code of
those specific tests.

No changes to Bitcoin Core are made - just unit tests added. The
Boost Test framework file `boost/test/execution_monitor.hpp` was
added to the lint whitelist - this is acceptable because it is a
documented public header of the Boost Test framework, which is
already part of the project.

- [x] `EvalChecksigTapscript`
- [x] `HandleMissingData`
- [x] `SignatureHashSchnorr` - now with full _path_ coverage
- [x] `GenericTransactionSignatureChecker<T>::VerifySchnorrSignature`
- [x] `GenericTransactionSignatureChecker<T>::CheckSchnorrSignature`
- [x] `ComputeTaprootMerkleRoot`
- [ ] `SigVersion::TAPSCRIPT` code paths in `ExecuteWitnessScript`
- [x] Taproot code paths in `VerifyWitnessProgram`
- [ ] Tapscript code paths in `VerifyWitnessProgram`

**TODO:** Full Tapscript (BIP-342) unit tests yet to be completed.
**TODO:** Does anything need to be done with `libconsensus`? Please advise!

**N.B.:** These unit tests include "death tests" (name comes from
Googletest framework): Tests that check that an `assert` actually
fails.  These are handled with a Boost `execution_monitor`
(mentioned above) but the asserts _still issue messages_ even when
by correctly trapped by the Boost Test framework.  You'll see the following:

```
gitpod /workspace/bitcoin/src/test (23279-taproot-unit-tests) $ ./test_bitcoin
Running 527 test cases...
test_bitcoin: script/interpreter.cpp:1495: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `false' failed.
test_bitcoin: script/interpreter.cpp:1497: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `in_pos < tx_to.vin.size()' failed.
test_bitcoin: script/interpreter.cpp:1528: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_annex_init' failed.
test_bitcoin: script/interpreter.cpp:1556: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_tapleaf_hash_init' failed.
test_bitcoin: script/interpreter.cpp:1559: bool SignatureHashSchnorr(uint256 &, ScriptExecutionData &, const T &, uint32_t, uint8_t, SigVersion, const PrecomputedTransactionData &, MissingDataBehavior) [T = CMutableTransaction]: Assertion `execdata.m_codeseparator_pos_init' failed.
test_bitcoin: script/interpreter.cpp:1469: bool HandleMissingData(MissingDataBehavior): Assertion `!"Missing data"' failed.
test_bitcoin: script/interpreter.cpp:1474: bool HandleMissingData(MissingDataBehavior): Assertion `!"Unknown MissingDataBehavior value"' failed.

*** No errors detected
gitpod /workspace/bitcoin/src/test (23279-taproot-unit-tests) $
```

These specific `Assertion ... failed` messages are _expected_.  The
tests triggering them properly _pass_ iff the `assert` (and no other)
is _triggered_.

david-bakin commented:
DrahtBot said:
> ### Conflicts
> Reviewers, this pull request conflicts with the following ones:
> 
> * [#22793](https://github.com/bitcoin/bitcoin/pull/22793) (Simplify BaseSignatureChecker virtual functions and GenericTransactionSignatureChecker constructors by achow101)

Referenced PR updates `BaseSignatureChecker` and `GenericSignatureChecker` which are used in these tests, and also various the internals of some routines in `interpreter.cpp` including `CheckSchnorrSignature`.

Most of the resulting conflicts are easy to resolve; the tricky one was the update to `CheckSchnorrSignature` which changed its behavior in such a way that one of these tests became invalid.
- This was exactly the case I was referring to in the PR header comment when I said:
  
  > somewhat fragile as they depend on code paths in the caller, which may change in the future. However, such tests are written so they will fail if the unit-under-test isn't actually called. This is noted in the code of those specific tests.

  And naturally it was immediately hit! But it was also easily fixable.

So in fact I am ready to update this PR if #22793 lands first.  And also, _except_ for that, I can update this code _now_ so that there won't be any conflict if #22793 lands later.

jamesob commented on src/test/util/pretty_data.cpp:
```diff
@@ -156,23 +153,22 @@ std::string FormatScriptFlags(unsigned int flags)
     for (const auto& se : script_errors)
         if (se.name == name)
             return se.err;
-    if (issue_boost_error) BOOST_ERROR("Unknown scripterror \"" << name << "\" in test description");
```
https://github.com/bitcoin/bitcoin/pull/25097/commits/eeefec343592c6de85048e3b055ba49755cfb072

You may want to move these `BOOST_ERROR(...)` changes back to the previous commit (where this file is introduced), since attempting to build the previous commit fails with this error:
```
/usr/bin/ld: /usr/bin/ld: DWARF error: could not find variable specification at offset 7218
libtest_util.a(libtest_util_a-pretty_data.o): in function `ParseScriptFlags(std::basic_string_view<char, std::char_traits<char> >, bool)':
./src/test/util/pretty_data.cpp:138: undefined reference to `boost::unit_test::unit_test_log_t::set_checkpoint(boost::unit_test::basic_cstring<char const>, unsigned long, boost::unit_test::basic_cstring<char const>)'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `lazy_ostream_impl':
/usr/include/boost/test/utils/lazy_ostream.hpp:74: undefined reference to `boost::unit_test::lazy_ostream::inst'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `ParseScriptFlags(std::basic_string_view<char, std::char_traits<char> >, bool)':
./src/test/util/pretty_data.cpp:138: undefined reference to `boost::test_tools::tt_detail::report_assertion(boost::test_tools::assertion_result const&, boost::unit_test::lazy_ostream const&, boost::unit_test::basic_cstring<char const>, unsigned long, boost::test_tools::tt_detail::tool_level, boost::test_tools::tt_detail::check_type, unsigned long, ...)'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `FormatScriptError[abi:cxx11](ScriptError_t, bool)':
./src/test/util/pretty_data.cpp:167: undefined reference to `boost::unit_test::unit_test_log_t::set_checkpoint(boost::unit_test::basic_cstring<char const>, unsigned long, boost::unit_test::basic_cstring<char const>)'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `lazy_ostream_impl':
/usr/include/boost/test/utils/lazy_ostream.hpp:74: undefined reference to `boost::unit_test::lazy_ostream::inst'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `FormatScriptError[abi:cxx11](ScriptError_t, bool)':
./src/test/util/pretty_data.cpp:167: undefined reference to `boost::test_tools::tt_detail::report_assertion(boost::test_tools::assertion_result const&, boost::unit_test::lazy_ostream const&, boost::unit_test::basic_cstring<char const>, unsigned long, boost::test_tools::tt_detail::tool_level, boost::test_tools::tt_detail::check_type, unsigned long, ...)'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `ParseScriptError(std::basic_string_view<char, std::char_traits<char> >, bool)':
./src/test/util/pretty_data.cpp:176: undefined reference to `boost::unit_test::unit_test_log_t::set_checkpoint(boost::unit_test::basic_cstring<char const>, unsigned long, boost::unit_test::basic_cstring<char const>)'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `lazy_ostream_impl':
/usr/include/boost/test/utils/lazy_ostream.hpp:74: undefined reference to `boost::unit_test::lazy_ostream::inst'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `ParseScriptError(std::basic_string_view<char, std::char_traits<char> >, bool)':
./src/test/util/pretty_data.cpp:176: undefined reference to `boost::test_tools::tt_detail::report_assertion(boost::test_tools::assertion_result const&, boost::unit_test::lazy_ostream const&, boost::unit_test::basic_cstring<char const>, unsigned long, boost::test_tools::tt_detail::tool_level, boost::test_tools::tt_detail::check_type, unsigned long, ...)'
/usr/bin/ld: libtest_util.a(libtest_util_a-pretty_data.o): in function `_GLOBAL__sub_I_pretty_data.cpp':
/usr/include/boost/test/unit_test_log.hpp:227: undefined reference to `boost::unit_test::unit_test_log_t::instance()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

jamesob commented on src/test/test_util_tests.cpp:
```diff
@@ -0,0 +1,308 @@
+
+#define EVAL(...) []() -> bool { return __VA_ARGS__; }()
+
+BOOST_AUTO_TEST_CASE(is_base_of_trait)
```
https://github.com/bitcoin/bitcoin/pull/25097/commits/12ae33708ee1978e2a215954e41e6ccf4f0a25fd

Is there a reason to include this test case? This doesn't look like it's testing any non-std functionality. It also seems like a direct adaptation of the example usage here: https://en.cppreference.com/w/cpp/types/is_base_of

jamesob reviewed (commented):
Partial review done

Code is high quality so far (if not a little complex), and I really like the look of these tests. Will finish review early next week, probably Monday. 

---


- [x] 12ae33708e Add new test utility methods; break out some existing ones

A lot of this code is fairly elaborate in terms of use of C++ templating
features - for example, all the machinations necessary to get `Hex()` to work -
which I think I'd be hesitant to include in non-test code. But since these are
tests, the complexity/convenience trade-off seems okay.

- [ ] eeefec3435 Unit tests for Tap{root,script} of `interpreter.cpp`, #23279

david-bakin commented:
> Partial review done

Addressing these comments now.

> 
> Code is high quality so far (if not a little complex), and I really like the look of these tests. Will finish review early next week, probably Monday.

Thank you, and looking forward to more comments!

Is it preferred (i.e., do you prefer) squashing _during_ a review or should I fix things in an additional commit on top of this one while you're in progress and squash it later?

jamesob commented:
> Is it preferred (i.e., do you prefer) squashing during a review or should I fix things in an additional commit on top of this one while you're in progress and squash it later?

Typically squashes are done immediately because if you squash at some point later, all reviewers will have to re-review to ensure that the code hasn't actually changed unexpectedly during the squash.

jamesob commented on src/test/script_tapscript_tests.cpp:
```diff
@@ -0,0 +1,1711 @@
+        Context& CheckSignatureCheckerWasCalled()
+        {
+            BOOST_CHECK_MESSAGE(m_checker_was_called,
+                                Descr() << ": Schnoor signature checker was called, as expected");
```
https://github.com/bitcoin/bitcoin/pull/25097/commits/eeefec343592c6de85048e3b055ba49755cfb072

"Schnoor" misspelling here and elsewhere

jamesob reviewed (commented):
approach ACK eeefec343592c6de85048e3b055ba49755cfb072 ([`jamesob/ackr/25097.1.david-bakin.test_unit_tests_for_tapr`](https://github.com/jamesob/bitcoin/tree/ackr/25097.1.david-bakin.test_unit_tests_for_tapr))

Wow, thanks for this contribution @david-bakin. These are some really
fine-grained tests for various Taproot related parts of the script interpreter.
The code here is very readable and seems nicely extensible. Clearly a
lot of work went into this.

For many parts of the codebase, this sort of testing would be overkill IMO
because it is so tightly coupled to the tested code. But in the case of script
interpretation - debatably the heart of bitcoin validation - I think it's
worthwhile to have painstakingly involved tests at the cost of slight
maintenance burden if the underlying code happens to change somehow. Maybe
someone more experienced than me w.r.t. changing the script interpreter can
weigh in here and confirm my intuition.

All of my comments are pretty minor. I'd say it might be wortwhile to break
out the `Parse*()` `optional` changes in the second commit either into their own
commit, or as I mentioned in a previous review, fold it into the first.

david-bakin commented:
yes sorry - i will update by tomorrow!


On Wed, Oct 12, 2022 at 12:21 PM Andrew Chow ***@***.***>
wrote:

> Are you still working on this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/25097#issuecomment-1276634618>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7YLBE6IOKK4MJDCZ47JYDWC4FUPANCNFSM5VOZLXLQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

achow101 commented:
> yes sorry - i will update by tomorrow!

It's been several weeks since this comment with no additional updates, so I'm going to close this for now. If you do get around to working on this, please leave a comment so that it can be reopened.

jamesob commented:
Sad to see this closed as it contains some novel unittest coverage. Please feel free to reopen with updates @david-bakin. Hopefully others will step in to review as well.

david-bakin commented:
Soon! Tied up temporarily

On Thu, Nov 10, 2022, 12:12 jamesob ***@***.***> wrote:

> Sad to see this closed as it contains some novel unittest coverage. Please
> feel free to reopen with updates @david-bakin
> <https://github.com/david-bakin>. Hopefully others will step in to review
> as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/25097#issuecomment-1310840025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7YLBHXF2WHJN3HJ2HN4E3WHVJKNANCNFSM5VOZLXLQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
